app/home/controllers.test.js

```diff
@@ -43,30 +43,23 @@ describe('Home controllers', function () {
 
       it('should return locations sorted by title', function () {
         const params = res.render.args[0][1]
-        expect(params).to.deep.equal({
-          pageTitle: 'dashboard::page_title',
-          sections: {
-            outgoing: {
-              permission: 'moves:view:outgoing',
-              heading: 'dashboard::sections.outgoing.heading',
-              filter: req.filterOutgoing,
-            },
-            incoming: {
-              permission: 'moves:view:incoming',
-              heading: 'dashboard::sections.incoming.heading',
-              filter: req.filterIncoming,
-            },
-            singleRequests: {
-              permission: 'moves:view:proposed',
-              heading: 'dashboard::sections.single_requests.heading',
-              filter: req.filterSingleRequests,
-            },
-            allocations: {
-              permission: 'allocations:view',
-              heading: 'dashboard::sections.allocations.heading',
-              filter: req.filterAllocations,
-            },
-          },
+        const sections = [
+          'outgoing',
+          'incoming',
+          'singleRequests',
+          'allocations',
+        ]
+        expect(params).to.have.all.keys('pageTitle', 'sections')
+        expect(params.pageTitle).to.equal('dashboard::page_title')
+        expect(params.sections).to.have.all.keys(sections)
+        sections.forEach(section => {
+          expect(params.sections[section]).to.have.all.keys([
```

Comment:
I notice this refactor now removes that we were testing that the sections contained the correct values for things like permissions and the filter. 

Interested on the thoughts of this approach, do you think it’s safe to not test this kind of behaviour in unit tests? I’m often in two minds...testing to means updating tests if the permission changes. But if not, it makes it easier to break when just updating the code. Sometimes a test explains the reason behind a value and can alert you if you change something you shouldn’t have.

Reply:
I don't think there's an easy answer here.
I usually strive to test the intent of the code. Testing that I'm using the same key twice (one in the code, one in the test) adds limited value.
As an example, if I want to refactor the code (perhaps renaming a key), I need to refactor the tests too because there is tight coupling with the rest of the code.
If the intent is to test the _shape_ of the objects (does the object look like a section?), I think there are better ways to do it with stronger guarantees (examples here are libraries such as [tcomb](https://github.com/gcanti/tcomb) or precompilers/linters such as Typescript).
Said that I have preferences, but I don't mind aligning with the rest of the codebase. If the tests above are too loose, I'm happy to revert to a stricter check of the keys/values.

Reply:
>As an example, if I want to refactor the code (perhaps renaming a key), I need to refactor the tests too because there is tight coupling with the rest of the code.

I think that's an interesting point. My view would be that if we didn't test the presence of particular keys to make refactoring easier it could open the door to regression. For example, if a view depends on an object having a particular key structure. If we don't test those keys (to make refactoring easier) that refactor could remove important keys and we wouldn't catch it.

I think my concern in this instance is how can we ensure that the filter value is correctly set to the value from the request object. This refactor removes the test that ensures that the filter key in this object is set from the request and if in a future change somebody no longer passes that value to this object, how would they know the expected behaviour has changed?

We could for example rely on the e2e tests picking that up but that also feels a bit slower and more fragile.